Fixed sized edge for any number of vertices

olafbooij opened the pull request:
Hi all,
I'm working on a hobby-slam-project which involves auto-calibration and thus requires a lot of edges which link 3 (or more) vertices, including a camera-pose-vertex, a landmark-vertex and a calibration-vertex. At the moment g2o can benefit from the compile-time information of the size of the vertices by using BaseUnaryEdge and BaseBinaryEdge. However, this is not possible for edges with more than 2 vertices.

I added a new edge type, BaseFixedSizedEdge, that can link any number of vertices specified at compile-time, using variadic templates. 
~~This new edge uses fold expressions available since c++17, which prohibits merging it into master.~~
This new edge uses some complex code which might be difficult to comprehend. It might therefore not be suitable for merging into master.
Nevertheless, it would perhaps be interesting as a new branch in g2o. Otherwise, I'm happy to maintain the fork (regularly pulling changes from g2o-master) if there's interest.
Note, if a merge with master is considered, then the docs should be updated.

jumonatr commented on CMakeLists.txt:
```diff
@@ -344,8 +344,9 @@ if(MSVC)
-set(CMAKE_CXX_STANDARD 11)
+# C++17 support
+#set(CMAKE_CXX_STANDARD 17) // CMAKE version >= 3.8
+set(CMAKE_CXX_FLAGS "-std=c++17")
```
You shouldn't switch back to the cmake_cxx_flag as some compilers need you to specify that differently (like msvc needs /std:c++ and I'm pretty sure gcc and clang are different.

olafbooij replied:
Agreed. I was reluctant to force a newer CMAKE version (I'm developing on a debian stable sytsem with comes with CMAKE 3.7.2). Fixed it.

olafbooij replied:
Fixed this by making the code C++11 compliant using your comments.

jumonatr commented on g2o/core/base_constant_edge.h:
```diff
@@ -0,0 +1,185 @@
+    }
+    constexpr std::pair<int, int> index_to_pair(const int k)
+    {
+      const int j = std::sqrt(2*(k+1))+.5;
```
std::sqrt isn't marked as constexpr in the standard so I don't think this works (at least not in clang or msvc)

olafbooij replied:
Good one, thanks! For now I'll add a constexpr sqrt impl. Though I'm also thinking an actual recursive function would perhaps be nicer...

olafbooij replied:
Added it to g2o/stuff/misc.h

jumonatr commented on g2o/core/base_constant_edge.h:
```diff
@@ -0,0 +1,185 @@
+
+namespace g2o {
+
+  namespace {
```
an anonymous namespace is redundant in a header, especially since constexpr functions are implicitly inline. If it weren't then all you'd be doing is adding these definitions in each of your translation units which I don't think is what you want to do here.

jumonatr commented on g2o/core/base_constant_edge.h:
```diff
@@ -0,0 +1,185 @@
+    using NthType = typename std::tuple_element<N, std::tuple<Types...>>::type;
+
+    template<typename Value, typename... Ts2>
+    auto tuple_init(const Value& value, const std::tuple<Ts2...>)
```
Since you're just using this to assign all the values in your tuple to nullptr it's redundant. All you need is to invoke the default constructor {}, and that will do it for you. You could even just initialize it at the point of declaration which would simplify this as well.

jumonatr replied:
Ah I see these aren't pointers you're trying to call the nullptr constructor of that type. It's annoying that they aren't default constructible that would have simplified things.

jumonatr replied:
If you defined a NullType you could do it like this:

template<typename Type>
using NullType = nullptr_t;

      BaseConstantEdge() : BaseEdge<D,E>(),
      _hessianRowMajor(false),
      _hessianTuple{NullType<VertexTypes>{}...},
      _hessianTupleTransposed{NullType<VertexTypes>{}...},
      _jacobianOplus({nullptr, D, VertexTypes::Dimension}...)
      {
        _vertices.resize(_nr_of_vertices, nullptr);
      }

olafbooij replied:
Ah, nice one. However gcc 6.3 gives strange compilation errors when I changed it like that. It starts complaining about the _jacobianOplus initialization, though that line is not changed. Do you know why this is?
<hmmm, I can't seem to attach a log of the error, will try later>

jumonatr replied:
Ah it's because in the other version you're expanding to nr_of_vertex_pairs not VertexTypes, I'll have to look it over again

olafbooij replied:
Ah, of course, I overlooked that. Well, by moving the trick inside struct HessianTupleType I can make it to work:
```
      struct HessianTupleType<std::integer_sequence<std::size_t, Ints...>>
      { 
        using type = std::tuple<HessianBlockTypeK<Ints>...>;
        using typeTransposed = std::tuple<HessianBlockTypeKTransposed<Ints>...>;
        template<std::size_t I>
        using NullType = nullptr_t; 
        static auto init() { return type{NullType<Ints>{}...};}
        static auto initTransposed() { return typeTransposed{NullType<Ints>{}...};}
      };
      using HessianTupleTypeH = HessianTupleType<std::make_index_sequence<_nr_of_vertex_pairs>>;
      BaseConstantEdge() : BaseEdge<D,E>(),
      _hessianRowMajor(false),
      _hessianTuple(HessianTupleTypeH::init()),
      _hessianTupleTransposed(HessianTupleTypeH::initTransposed()),
      _jacobianOplus({nullptr, D, VertexTypes::Dimension}...)
```
But that looks actually more complicated than the tuple_init function...

jumonatr replied:
Yeah I think you're right, do what you think is best. You may want to change that tuple init not to grab that tuple by value (const&). Or just use tuple_init<decltype(_hessionTuple)> instead of the parameter to do the capture of the tuple type.

olafbooij replied:
Changed it into a const&.

jumonatr commented on g2o/core/base_constant_edge.h:
```diff
@@ -0,0 +1,185 @@
+
+      BaseConstantEdge() : BaseEdge<D,E>(),
+      _hessianRowMajor(false),
+      _hessianTuple(tuple_init(nullptr, _hessianTuple)),
```
your tuple_init is equivalent to as your just trying to null them out:

_hessianTuple{},
_hessianTupleTransposes{}

olafbooij replied:
(as you round out above, this is not possible because of Eigen's choice not to have them default constructable.)

jumonatr commented on g2o/core/base_constant_edge.h:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    template<int N, typename... Types>
+    using NthType = typename std::tuple_element<N, std::tuple<Types...>>::type;
```
Since you're just using this once, you might just want to define it in your class so you're not adding this name is the g2o namespace (same goes for most of these other helpers)

olafbooij replied:
Agreed, I moved it in.

jumonatr commented on g2o/core/base_constant_edge.hpp:
```diff
@@ -0,0 +1,271 @@
+template <int D, typename E, typename... VertexTypes>
+void BaseConstantEdge<D, E, VertexTypes...>::mapHessianMemory(number_t* d, int i, int j, bool rowMajor)
+{
+  auto f = [d](auto& hessian){ new (&hessian) std::remove_reference_t<decltype(hessian)>(d); };
```
calling a placement new on a non-trivial type is really sketchy as it won't destroy the object that was there beforehand. You should just call a regular assignment operator/constructor.

olafbooij replied:
I copied here the logic that g2o uses in all its edge classes. I do not know exactly why this choice was made. It would be good first to check this first, and perhaps change it in a different PR, also involving the dynamic base_multi_edge.

jumonatr commented on g2o/core/base_constant_edge.hpp:
```diff
@@ -0,0 +1,271 @@
+template <int... Ints>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplusNs(std::integer_sequence<int, Ints...>)
+{
+  ( void(linearizeOplusN<Ints>()), ...);
```
You could just require C++11 if instead of using this you just used regular expansion using this trick:

int unused[] = { (linearizeOPlusN<Ints>(), 0)… };
(void)unused;

olafbooij replied:
Would that work for all the fold expressions I used?
If so, then that would be great, because it would mean we can drop the dependency on c++17.

jumonatr replied:
Yeah pretty much all your folds can be replaced, like for example:

Even the ones that use the results can be changed like so:
` 
bool fixed[] = {
    vertexXn<Ints>()->fixed()...
};
return std::all_of(std::begin(fixed), std::end(fixed), [](bool value){ return value; });
`

olafbooij replied:
I followed your advise and changed all code to make it c++11 compliant.

jumonatr commented on g2o/core/base_unary_edge.h:
```diff
@@ -27,75 +27,20 @@
+	// _jacobianOplusXi is used.
   template <int D, typename E, typename VertexXi>
-  class BaseUnaryEdge : public BaseEdge<D,E>
+  class BaseUnaryEdge : public BaseConstantEdge<D, E, VertexXi>
```
BaseBinaryEdge and BaseUnaryEdge are now pretty useless, this change would allow us to just remove them.

olafbooij replied:
Indeed. However, they do now seem to be a part of the API and are used extensively in the types, the examples, and most probably in other projects that depend on g2o. I'm reluctant to break those.

jumonatr commented on g2o/core/base_constant_edge.hpp:
```diff
@@ -0,0 +1,271 @@
+};
+
+template<>
+struct Tuple_apply_i<0>
```
you may want to consider putting these helpers (like the ones in base_constant_edge.h as well) in a stuff/traits.h file or something instead of putting them all inline here

olafbooij replied:
I moved them to a new file g2o/stuff/tuple_tools.h and added some unit-tests in unit_test/stuff/tuple_tools_tests.cpp.

jumonatr commented:
I like the idea of this change as it removes some duplication, but I think enforcing C++17 might be a little much right now. I'm pretty sure we can massage this change into something that would just require C++11 (or C++14) with a little elbow grease. It's also worth noting that it does add quite a bit of complexity (although I think we can simplify a lot of what is happening here to make it more intuitive) so I don't know if it's worth it as it also makes understanding what's going on under the hood quite a bit more difficult for people that aren't well versed in templates. I don't have a strong opinion either way, I just wanted to put this out there as a talking point.

sjulier commented:
Would it be possible to consider if the names are right here? To me, "const" implies "value does not change" (e.g., similar to a fixed vertex). However, the actual situation we have here is a parallel between std::vector and std::array.

The C++ standard describes std::array as being a "fixed sized array". Therefore, would something which reflects this - perhaps "BaseFixedSizedEdge" - be more appropriate instead? MultiEdge could then be renamed as "BaseVariableSizedEdge".

olafbooij commented:
I followed Simon Julier's advise and changed the new edge class to BaseFixedSizedEdge and, in addition, the BaseMultiEdge to BaseVariableSizedEdge. I added an alias for BaseMultiEdge for backward-compatibility.

olafbooij commented:
The tests are still failing I see. I managed to reproduce this using Docker. It seems to be happening for gcc 7.3 as used in the tests and not for gcc 6.3 which I'm using for development (nor gcc 8.0 btw).

olafbooij commented:
To not clutter the project I closed this PR. It will live on as a branch in the forked repository.

YangSiri commented:
@olafbooij  Hi, so is there any tutorial for mulitedge of dynamic number of vertexs?

RainerKuemmerle commented:
Now as a branch on here, also updated it on existing master
